elrond-wasm/src/storage/mappers/bi_di_mapper.rs

```diff
@@ -0,0 +1,269 @@
+    id_set_mapper: UnorderedSetMapper<SA, K>,
+    value_set_mapper: UnorderedSetMapper<SA, V>,
+    base_key: StorageKey<SA>,
+}
+
+impl<SA, K, V> StorageMapper<SA> for BiDiMapper<SA, K, V>
+where
+    SA: StorageMapperApi,
+    K: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+    V: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+{
+    fn new(base_key: StorageKey<SA>) -> Self {
+        let mut id_key = base_key.clone();
+        id_key.append_bytes(ID_SUFIX);
+
+        let mut value_key = base_key.clone();
+        value_key.append_bytes(VALUE_SUFIX);
+        BiDiMapper {
+            _phantom_api: PhantomData,
+            id_set_mapper: UnorderedSetMapper::<SA, K>::new(id_key),
+            value_set_mapper: UnorderedSetMapper::<SA, V>::new(value_key),
+            base_key,
+        }
+    }
+}
+
+impl<SA, K, V> BiDiMapper<SA, K, V>
+where
+    SA: StorageMapperApi,
+    K: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+    V: TopEncode
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+{
+    fn get_id_key(&self, value: &V) -> StorageKey<SA> {
+        let mut key = self.base_key.clone();
+        key.append_bytes(VALUE_TO_ID_SUFFIX);
+        key.append_item(value);
+        key
+    }
+
+    fn get_value_key(&self, key: &K) -> StorageKey<SA> {
+        let mut value = self.base_key.clone();
+        value.append_bytes(ID_TO_VALUE_SUFFIX);
+        value.append_item(&key);
+        value
+    }
+
+    pub fn get_id(&self, value: &V) -> K {
+        storage_get(self.get_id_key(value).as_ref())
+    }
+
+    fn set_id(&mut self, value: &V, id: &K) {
+        storage_set(self.get_id_key(value).as_ref(), id);
+    }
+
+    pub fn get_value(&self, id: &K) -> V {
+        storage_get(self.get_value_key(id).as_ref())
+    }
+
+    fn set_value(&mut self, id: &K, value: &V) {
+        storage_set(self.get_value_key(id).as_ref(), value);
+    }
+
+    pub fn insert(&mut self, id: K, value: V) -> bool {
+        if self.id_set_mapper.contains(&id) || self.value_set_mapper.contains(&value) {
+            return false;
+        }
+        self.set_id(&value, &id);
+        self.set_value(&id, &value);
+
+        self.id_set_mapper.insert(id);
+        self.value_set_mapper.insert(value);
+        true
+    }
+
+    pub fn remove_by_id(&mut self, id: &K) -> bool {
+        if self.id_set_mapper.swap_remove(id) {
+            let value = self.get_value(id);
+            storage_clear(self.get_id_key(&value).as_ref());
+            storage_clear(self.get_value_key(id).as_ref());
```

Comment:
Extract those into private methods, something like `clear_id()` and `clear_value()` respectively.